### DISCUSSION TITLE
Difference between AsyncSubject and ResourceSubject and Shaders.Input and Windows.Input

### DISCUSSION THREAD

**Discussion by aniachr3** [bonsai-rx/bonsai#1562] [Q&A]
Hi all, 

as in the title, I've been struggling to understand what is the difference between:
- AsyncSubject and ResourceSubject;
- KeyDown (Windows.Input) vs. KeyDown (Shaders.Input).

Thanks in advance!

**Comment by bruno-f-cruz** [comment]
Heya,
Have you checked the [docs](https://bonsai-rx.org/docs/articles/subjects.html)?
Regarding the second question, they use different APIs. Windows.Input uses the OS API to record any keyboard input, Shaders uses the OpenTK package and will only get values if the shaders' window is interacted with. They also have different temporal precision since the KeyDown (win) is tied to a OS event while the shader's one is tied to the shader's scheduler so it will emit values on a `UpdateFrame` event.

**Reply by aniachr3** [reply]
Hi Bruno,
yes, I checked the docs, but the description and marble diagrams are basically identical for both, except for the Warning about IDisposable stored value in the ResourceSubject section, which is not entirely clear to me.
Thanks for the explanation of Windows vs. Shaders!

**Reply by bruno-f-cruz** [reply]
`ResourceSubjects` dispose of the object when the scope of the subject is terminated. This is really nice when you are dealing with disposable resources like device objects. For the most part, however, bonsai already takes care of these patterns for you, but in the future, we might want to leverage this type of subject more and more. For now, they are used for a bit more niche applications. All the other subjects are not able to call the `Dispose` method. So they are better suited for non-disposable objects.
